### PR TITLE
[POC] In a world where job events have no counts...

### PR DESCRIPTION
**Pull request by jakemcdermott**
Sending the total count for the job events list is expensive from a backend performance perspective. In this PR, we'll try to envision how or if the ui might work without it.

Currently, the total `count` determines the scroll range of our job output ui. It is also how we know where the current "end" of a running job is for when the user decides to skip to the bottom of the job results using the `page last` button.

#### Jobs that are still running
... we'd probably make additional api requests that sort the `job_events` list by `-counter` on initial page load and whenever the user clicks `page last`. We'd use the `counter` from the first event in the results to scale the scroll range or parametrize the `last` page request. See the file changes in this PR for a sketch of how that might work technically.

#### Jobs that are finished
I'm unsure about how we might support our current navigation controls without a `count` at the moment and am very much open to your thoughts and suggestions. 

For finished jobs, users can apply search filters, scroll through the paginated search results, and skip to the last page. **[1]** When a search filter is applied, the highest job event counter can't be used to establish the scroll range because event counters apply to the whole collection of events and not the filtered set of events.

**[1]** [Mockups](https://tower-mockups.testing.ansible.com/patternfly/jobs/jobs-output/)

Features:
   - scrolling and page navigation controls for job results that have search filters applied

### CODE REVIEW

File: awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx [hunk 1]
```diff
@@ -269,9 +282,18 @@ class JobOutput extends Component {
...
+  async handleScrollLast() {
+    const { job, type } = this.props;
+    const {
+      data: {
+        results: [{ counter }],
+      },
+    } = await JobsAPI.readEvents(job.id, type, {
+      page_size: 1,
+      order_by: '-counter',
+    });
```
Comment (jakemcdermott) [inline awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx:294]: **UX considerations:**

Given a job with 40,000 events on our internal test instance, this seems to add about ~2s to the total loading time when clicking `page last` (on average). 

Screenshot from 2020-01-16 12-04-04

Note: our internal test instance doesn't include some of our recent api-side optimizations for jobs and job events.

File: awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx [hunk 2]
```diff
@@ -269,9 +282,18 @@ class JobOutput extends Component {
...
+    this.setState({ remoteRowCount: counter + 1 });
+    this.scrollToRow(counter);
   }
```
Comment (jakemcdermott) [inline awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx:297]: @ryanpetrello @trahman73 @wenottingham 

One path forward I can see for making this work without a total count is to _turn off pagination scrolling_ in the job output viewer when a search filter is applied. 

Applying a search filter to the results would put the output ui into a new mode. In this mode, only the `up`, `down`, `first`, and `last` navigation buttons would be available. There would be no infinite scrolling of filtered job events.

Some drawbacks here that I can think of:
  - If the filtered results set is huge, there may be a lot of button clicking in order to navigate the results.
  - Having the ability to scroll in some states but not others could make for a strange UX
  - adds some complexity

Comment (jakemcdermott) [reply awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx:297]: @ryanpetrello @trahman73 @wenottingham 

One path forward that I can see for making this feature work without a total count is to handle scrolling and navigation differently when a search filter is applied. 

**unfiltered job results and running jobs:** The page would have a virtualized / "infinite" scroll with a range that is adjusted by the total event count, just as it is now in ui_next. We can use the highest event counter as the total count in this scenario.

**jobs with search filter applied:** Scrolling would have to work a little more like it does in current awx. The scroll range will need to grow as you move up or down into new regions, and it'll completely reset if you do a `first` or `last` navigation.

**Comment by jlmitch5** [comment]
> For finished jobs, users can apply search filters, scroll through the paginated search results, and skip to the last page. **[1]** When a search filter is applied, the highest job event counter can't be used to establish the scroll range because event counters apply to the whole collection of events and not the filtered set of events.

This to me makes me feel like we need the count.

I was thinking if the issue is that it is expensive for the API to "compute" the count, then I feel like a solution could be to make it so you don't compute it on read (I'm assuming the way we are getting this is to find the number of records in the events "table"), but rather have it as a number of events data item that we store and can access.  So each time a new event is added the "count" item increments.  If the issue is "count" is a reserved field, we could have it be total_count or something.  Unfortunately, this value would only be relevant for _total_ count, not for filtered jobs.  But maybe it is enough to help improve performance by default?

**Comment by ryanpetrello** [comment]
@jakemcdermott if https://github.com/ansible/awx/pull/5710 pans out, I don't think we actually need this PR/change at all.

**Comment by jakemcdermott** [comment]
:confetti_ball: Closing this for now, see https://github.com/ansible/awx/pull/5710!

File: awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx (hunk 1 above)
Comment (ryanpetrello) [reply awx/ui_next/src/screens/Job/JobOutput/JobOutput.jsx:294]: This is slow because `counter` isn't indexed.

**Pull request closed without merging**